x/fee_grant/internal/types/periodic_fee.go

```diff
@@ -0,0 +1,129 @@
+// Keeper.UseGrantedFees and the return values should match how it is handled there.
+//
+// If it returns an error, the fee payment is rejected, otherwise it is accepted.
+// The FeeAllowance implementation is expected to update it's internal state
+// and will be saved again after an acceptance.
+//
+// If remove is true (regardless of the error), the FeeAllowance will be deleted from storage
+// (eg. when it is used up). (See call to RevokeFeeAllowance in Keeper.UseGrantedFees)
+func (a *PeriodicFeeAllowance) Accept(fee sdk.Coins, blockTime time.Time, blockHeight int64) (bool, error) {
+	if a.Basic.Expiration.IsExpired(blockTime, blockHeight) {
+		return true, ErrFeeLimitExpired()
+	}
+
+	a.TryResetPeriod(blockTime, blockHeight)
+
+	// deduct from both the current period and the max amount
+	var isNeg bool
+	a.PeriodCanSpend, isNeg = a.PeriodCanSpend.SafeSub(fee)
+	if isNeg {
+		return false, ErrFeeLimitExceeded()
+	}
+	a.Basic.SpendLimit, isNeg = a.Basic.SpendLimit.SafeSub(fee)
+	if isNeg {
+		return false, ErrFeeLimitExceeded()
+	}
+
+	return a.Basic.SpendLimit.IsZero(), nil
+}
+
+// TryResetPeriod will check if the PeriodReset has been hit. If not, it is a no-op.
+// If we hit the reset period, it will top up the PeriodCanSpend amount to
+// min(PeriodicSpendLimit, a.Basic.SpendLimit) so it is never more than the maximum allowed.
+// It will also update the PeriodReset. If we are within one Period, it will update from the
+// last PeriodReset (eg. if you always do one tx per day, it will always reset the same time)
+// If we are more then one period out (eg. no activity in a week), reset is one Period from the execution of this method
+func (a *PeriodicFeeAllowance) TryResetPeriod(blockTime time.Time, blockHeight int64) {
+	if !a.PeriodReset.IsZero() && !a.PeriodReset.IsExpired(blockTime, blockHeight) {
+		return
+	}
+	// set CanSpend to the lesser of PeriodSpendLimit and the TotalLimit
+	if _, isNeg := a.Basic.SpendLimit.SafeSub(a.PeriodSpendLimit); isNeg {
+		a.PeriodCanSpend = a.Basic.SpendLimit
+	} else {
+		a.PeriodCanSpend = a.PeriodSpendLimit
+	}
+
+	// If we are within the period, step from expiration (eg. if you always do one tx per day, it will always reset the same time)
+	// If we are more then one period out (eg. no activity in a week), reset is one period from this time
+	a.PeriodReset = a.PeriodReset.MustStep(a.Period)
+	if a.PeriodReset.IsExpired(blockTime, blockHeight) {
+		a.PeriodReset = a.PeriodReset.FastForward(blockTime, blockHeight).MustStep(a.Period)
+	}
+}
+
+// PrepareForExport will adjust the expiration based on export time. In particular,
+// it will subtract the dumpHeight from any height-based expiration to ensure that
+// the elapsed number of blocks this allowance is valid for is fixed.
+// (For PeriodReset and Basic.Expiration)
+func (a *PeriodicFeeAllowance) PrepareForExport(dumpTime time.Time, dumpHeight int64) exported.FeeAllowance {
+	return &PeriodicFeeAllowance{
```

Comment:
Do we need to create a custom marshaller for this for a better UX? Idk if you can query the allowances. I'm asking this bc we had a similar issue with #5270. cc @alexanderbez

Reply:
Doesn't go-amino json encoding work?

Reply:
It works, yes. Idk what fede is suggesting though. Custom JSON marshalling is only used when you want to flatten the top-level keys...cause Amino sucks.